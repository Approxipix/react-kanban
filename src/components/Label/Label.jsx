import React from 'react';
import styled from 'styled-components';

const LabelList = styled.ul`
  display: flex;
`;

const LabelItem = styled.li`
  width: 2rem;
  height: 2rem;
  margin-right: .5rem;
  background-color: ${props => props.value};
  border-radius: .2rem;
`;


const Label = (props) => {
  const { cardLabels } = props;
  if (!cardLabels) return null;
  return (
    <LabelList>
      {cardLabels.map((label, index) => (
        <LabelItem key={index} value={label.color}/>
      ))}
    </LabelList>
  )
};

export default Label;
