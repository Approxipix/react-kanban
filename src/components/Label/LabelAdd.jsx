import React, { Component } from "react";
import PropTypes from 'prop-types';
import { connect } from "react-redux";
import { bindActionCreators } from 'redux';
import { addLabelToCard } from "../../redux/cardReducer/actions/actions";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import styled from 'styled-components';
import Label from "./Label";

const LabelList = styled.ul`
  display: flex;
  flex-direction: column;
  padding: 0 .5rem;
`;

const LabelItem = styled.li`
  position: relative;
  height: 2.25rem;
  width: 100%;
  margin-bottom: .3rem;
  background-color: ${props => props.value};
  cursor: pointer;
  border-radius: .2rem;
  &:before {
    content: '';
    position: absolute;
    top: 0; 
    bottom: 0; 
    left: 0;
    width: .6rem;
    border-radius: .2rem 0 0 .2rem;
    background-color: ${props => props.value};
    box-shadow: inset 2rem 0 rgba(0, 0, 0, .2);
    transition: transform .2s ease-in-out;
    z-index: -1;
  }
  &:hover {
    &:before {
      transform: translateX(-90%);
    }
  }
`;

const LabelIcon = styled.div`
  color: #fff;
  position: absolute;
  top: 50%;
  left: 50%;
  font-size: 1rem;
  transform: translate(-50%, -50%);
`;

class LabelAdd extends Component {
  addLabel = (label) => {
    this.props.actions.addLabelToCard({
      cardId: this.props.cardId,
      cardLabel: label
    })
  };

  render() {
    const { cardLabels, labels } = this.props;
    if (!labels) return null;
    return (
      <LabelList>
        {labels.map((label, index) => (
          <LabelItem
            key={index}
            value={label.color}
            onClick={() => this.addLabel(label)}
          >
            {cardLabels && cardLabels.some(cardLabel => cardLabel._labelId === label._labelId) && (
              <LabelIcon>
                <FontAwesomeIcon icon="check" />
              </LabelIcon>
            )}
          </LabelItem>
        ))}
      </LabelList>
    )
  };
}

LabelAdd.propTypes = {
  labels: PropTypes.arrayOf(
    PropTypes.shape({
      _labelId: PropTypes.number.isRequired,
      color: PropTypes.string.isRequired
    })
  ),
  cardLabels: PropTypes.arrayOf(
    PropTypes.shape({
      _labelId: PropTypes.number.isRequired,
      color: PropTypes.string
    })
  ),
};

function mapStateToProps(state) {
  return {
    labels: state.rootReducer.labels,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      addLabelToCard: addLabelToCard,
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(LabelAdd);
