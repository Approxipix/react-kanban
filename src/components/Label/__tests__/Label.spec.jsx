import React from 'react'
import { shallow } from 'enzyme'
import toJson from "enzyme-to-json";
import Label from '../Label'

describe('<Label>', () => {
  let component;
  beforeEach(() => {
    const props = {
      cardLabels: [{color: '#fff'}, {color: '#ccc'}, {color: '#000'}]
    };
    component = shallow(<Label {...props} />);
  });

  it('should renders correctly without props', () => {
    const wrapper = shallow(<Label />);

    expect(toJson(wrapper)).toMatchSnapshot()
  });

  it('should renders without crashing given the required props', () => {
    expect(toJson(component)).toMatchSnapshot()
  });

  it('should render a Label item for each cardLabels', () => {
    expect(component.children().length).toEqual(3)
  });

  it('should sets the color prop as the `value` on the Label item', () => {
    expect(component.childAt(0).props().value).toEqual('#fff');
    expect(component.childAt(1).props().value).toEqual('#ccc');
    expect(component.childAt(2).props().value).toEqual('#000');
  });
});
