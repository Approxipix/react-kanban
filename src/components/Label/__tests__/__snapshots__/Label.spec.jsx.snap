// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Label> should renders correctly without props 1`] = `""`;

exports[`<Label> should renders without crashing given the required props 1`] = `
<styled.ul
  className="label-list"
>
  <styled.li
    key="0"
    value="#fff"
  />
  <styled.li
    key="1"
    value="#ccc"
  />
  <styled.li
    key="2"
    value="#000"
  />
</styled.ul>
`;
